src/main/java/com/gluonhq/richtext/RichTextAreaSkin.java

```diff
@@ -138,18 +142,40 @@ public void dispose() {
     //TODO Need more optimal way of rendering text fragments.
     //  For now rebuilding the whole text flow
     private void refreshTextFlow() {
-        var fragments = new ArrayList<Text>();
-        viewModel.walkFragments( (text, decoration) -> fragments.add( buildText(text, decoration)));
-        textFlow.getChildren().setAll(fragments);
+        fontCacheEvictionTimer.pause();
+        try {
+            var fragments = new ArrayList<Text>();
+            viewModel.walkFragments((text, decoration) -> fragments.add(buildText(text, decoration)));
+            textFlow.getChildren().setAll(fragments);
+        } finally {
+            fontCacheEvictionTimer.start();
+        }
     }
 
     private Text buildText(String content, TextDecoration decoration ) {
-        Text text = new Text(content);
-        Optional.ofNullable(decoration.getForeground()).ifPresent(text::setFill); // has to be fill for font to render properly
-        Optional.ofNullable(decoration.getFont()).ifPresentOrElse(text::setFont, () -> text.setFont(TextDecoration.DEFAULT.getFont()) );
+        Objects.requireNonNull(decoration);
+        Text text = new Text(Objects.requireNonNull(content));
+        text.setFill(decoration.getForeground());
+
+        // Cashing fonts, assuming their, especially for default one
+        String hash = decoration.getFontFamily() + decoration.getFontWeight() + decoration.getFontPosture() + decoration.getFontSize();
+        Font font = fontCache.computeIfAbsent( hash,
+            h -> Font.font( decoration.getFontFamily(), decoration.getFontWeight(), decoration.getFontPosture(), decoration.getFontSize()));
+
+        text.setFont(font);
         return text;
     }
 
+    private void evictUnusedFonts() {
+        List<Font> usedFonts =  textFlow.getChildren()
+                .stream()
+                .filter(n -> n instanceof Text)
+                .map( t -> ((Text)t).getFont()).collect(Collectors.toList());
```

Comment:
I don't think font could be null, but maybe add a filter to prevent it?

Reply:
Done